### ISSUE TITLE
birdseye is not fully closed upon exit of birdseye

### ISSUE THREAD

**Issue by Almenon** [Almenon/birdseye-vscode#11]
When the user closes birdseye I close the process I started.  This works.

The problem is that birdseye launches a second python process (I'm guessing the server) - I need to find some multi-platform way of closing that as well.

![image](https://user-images.githubusercontent.com/13080965/38166823-618d2d86-34df-11e8-8609-b76442b3ac30.png)

There should be some way of identifying child process - but I did not see any grandchildren of my process in the debugger.

![image](https://user-images.githubusercontent.com/13080965/38166797-e1b4155c-34de-11e8-963e-2ead74924753.png)

Or maybe there is a way of killing a process and all of its children all at once?

**Comment by Almenon** [comment]
https://stackoverflow.com/questions/33325301/killing-a-bash-script-does-not-kill-child-processes

This looks like it might work

**Comment by Almenon** [comment]
Nevermind - doesn't work for windows :/

Open a process with [detached](https://nodejs.org/api/child_process.html#child_process_child_process_spawn_command_args_options) on windows causes the process to pop open with a console window.
Trying to kill with a negative pid results in Error: kill ESRCH

Apparently killing all children is not supported on windows :(

https://github.com/nodejs/node/issues/3617

Or at least not with node - I could spin up a subprocess to nuke them in another language.

This guy did exactly that: http://krasimirtsonev.com/blog/article/Nodejs-managing-child-processes-starting-stopping-exec-spawn

**Comment by Almenon** [comment]
Fixed :D

```typescript
import {exec} from 'child_process'

/**
 * kills the process and all its children
 * If you are on linux process needs to be launched in detached state
 * @param pid process identifier
 * @param signal kill signal
 */
export function killAll(pid:number, signal:string|number='SIGKILL'){
    if(process.platform == "win32"){
        exec(`taskkill /PID ${pid} /T /F`, (error, stdout, stderr)=>{
            console.log("stdout: " + stdout)
            console.log("stdout: " + stderr)
            if(error){
                console.log("error: " + error.message)
            }
        })
    }
    else{
        // see https://nodejs.org/api/child_process.html#child_process_options_detached
        // If pid is less than -1, then sig is sent to every process in the process group whose ID is -pid.
        process.kill(-pid, signal)
    }
}
```

**Comment by Almenon** [comment]
Well forgot to test on ubuntu but I just tested on mac and it looks like it works

1. Launch birdseye

2. > ps -ax | grep python

3. Two results

4. Close birdseye

5. > ps -ax | grep python

6. no results.

@alexmojaki I published a medium article on how to kill processes with node - does the explanation make sense?  Any thoughts?

https://medium.com/@almenon214/killing-processes-with-node-772ffdd19aad

**Comment by alexmojaki** [comment]
It looks mostly fine. Some small points:

- You mention 'Karsimiov' but the linked blog says 'Krasimir'
- The title is not very descriptive, mention that you're killing a tree/family/group of processes
- The comment I made in https://github.com/Almenon/birdseye-vscode/commit/5280cd2a573322b46bdfe2d55f88c86d48c7c85e

**Comment by Almenon** [comment]
good catch.  Fixed typo made in 5280cd2 and the gist.  Updated the article to change the title and name.

https://medium.com/@almenon214/killing-processes-with-node-772ffdd19aad
